### PR TITLE
cobalt: Implement app lifecycle and window management

### PR DESCRIPTION
**Pull request by jellefoks**
Introduce the core infrastructure for handling application lifecycle events and enhance Starboard window management robustness.

With this commit, conceal, reveal, focus, and blur functionality is fully operational with respect to lazy Starboard window creation/destruction and the corresponding Web API focus and visibility states.

Key Changes:
- Implement static entry points in content::Shell for lifecycle events (Blur, Focus, Conceal, Reveal, Freeze, Unfreeze, Stop).
- Add virtual handlers to ShellPlatformDelegate to propagate these events.
- Implement explicit focus/blur propagation to WebContents in ShellPlatformDelegate (OnFocus calls WebContents::Focus, OnBlur calls RenderWidgetHost::Blur).
- Update AppLifecycleDelegate to bridge global Starboard events to the newly added Shell handlers.
- Enhance PlatformWindowStarboard to robustly manage window creation, destruction, and visibility transitions, preventing redundant events.
- Map SIGWINCH to the Blur event in the POSIX signal handler to support integrated lifecycle testing.
- Introduce test_lifecycle.sh and enhance test_preload.sh to verify Visibility and Focus API states using the Chrome DevTools Protocol (CDP).
- Fix cyclic dependencies in Starboard Ozone by removing redundant ui/aura and ui/compositor inclusions.
- Ensure proper widget activation and visibility management in DesktopWindowTreeHostPlatform for Starboard.
- Enhance AppLifecycleDelegateTest to verify WebContents focus state transitions.

All changes are verified with unit tests and automated integration scripts on linux-x64x11-modular.

Bug: 448156280

### CODE REVIEW

File: starboard/shared/signal/suspend_signals.cc [hunk 1]
```diff
@@ -65,6 +65,13 @@ void Conceal(int signal_id) {
   SignalMask(kAllSignals, SIG_UNBLOCK);
 }
 
+void Blur(int signal_id) {
+  SignalMask(kAllSignals, SIG_BLOCK);
+  LogSignalCaught(signal_id);
+  Application::Get()->Blur(NULL, NULL);
```
[the remark this answers is not in the document]
Comment (jellefoks) [reply starboard/shared/signal/suspend_signals.cc]: Thanks for the suggested cleanup. Done.

File: cobalt/app/app_lifecycle_delegate.cc [hunk 2]
```diff
@@ -144,7 +144,20 @@ void AppLifecycleDelegate::HandleEvent(const SbEvent* event) {
...
+      content::Shell::OnBlur();
+#if BUILDFLAG(IS_STARBOARD)
+      if (platform_event_source_) {
+        platform_event_source_->HandleFocusEvent(event);
+      }
+#else
+      // On Android, focus is handled natively via Activity and View lifecycle
+      // callbacks (e.g. CobaltActivity.onResume) which propagate directly to
+      // Chromium's WindowAndroid.
+      NOTIMPLEMENTED();
```
Comment (andrewsavage1) [inline cobalt/app/app_lifecycle_delegate.cc]: Should we call NOTIMPLEMENTED if it sounds like it is implemented elsewhere?

File: cobalt/shell/browser/shell_platform_delegate.cc [hunk 3]
```diff
@@ -25,16 +27,64 @@ bool ShellPlatformDelegate::IsVisible() const {
   return is_visible_;
 }
 
+__attribute__((weak)) void ShellPlatformDelegate::Initialize(
```
Comment (andrewsavage1) [inline cobalt/shell/browser/shell_platform_delegate.cc]: Why do we need the weak attribute? Can we remove it?

File: cobalt/shell/browser/shell_test_support.h [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2026 The Cobalt Authors. All Rights Reserved.
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
```
Comment (andrewsavage1) [inline cobalt/shell/browser/shell_test_support.h]: Change back to 2026

File: starboard/shared/x11/application_x11.cc [hunk 5]
```diff
@@ -897,10 +897,12 @@ Application::Event* ApplicationX11::WaitForSystemEventWithTimeout(
 }
 
 void ApplicationX11::WakeSystemEventWait() {
-  SB_DCHECK(!windows_.empty());
-  XSendAtom((*windows_.begin())->window, wake_up_atom_);
-  SB_DCHECK(dev_input_);
-  dev_input_->WakeSystemEventWait();
+  if (!windows_.empty()) {
```
Comment (andrewsavage1) [inline starboard/shared/x11/application_x11.cc]: Why does this function need to change? Will we have cases in Chrobalt (as opposed to Cobalt 25 and earlier) where these checks would fail?

**Comment by jellefoks** [comment]
Whoop! this is ready for review.

File: cobalt/app/app_lifecycle_delegate.cc [hunk 6]
```diff
@@ -144,7 +144,20 @@ void AppLifecycleDelegate::HandleEvent(const SbEvent* event) {
...
     case kSbEventTypeBlur:
+      content::Shell::OnBlur();
+#if BUILDFLAG(IS_STARBOARD)
+      if (platform_event_source_) {
+        platform_event_source_->HandleFocusEvent(event);
+      }
+#else
+      // On Android, focus is handled natively via Activity and View lifecycle
+      // callbacks (e.g. CobaltActivity.onResume) which propagate directly to
+      // Chromium's WindowAndroid.
```
Comment (yell0wd0g) [inline cobalt/app/app_lifecycle_delegate.cc]: nit: This `!IS_STARBOARD` branch would apply to tvOS as well, not 
just Android.

But looking at BUILD.gn, the associated unittest file seems to
only be built for starboard platforms ([code](https://source.corp.google.com/h/lbshell-internal/cobalt_src/+/26.eap:cobalt/BUILD.gn;l=201?q=app_lifecycle_delegate.cc%20&ss=h%2Flbshell-internal%2Fcobalt_src%2F%2B%2Frefs%2Fheads%2F26.eap:cobalt%2F)) and
a quick glance at an android-arm-qa build like [here](https://github.com/youtube/cobalt/actions/runs/22967686521/job/66675953118?pr=9423) doesn't list it, 
so this buildflag should rather be:
```
#if !BUILDFLAG(IS_STARBOARD)
#error This file is only intended for Starboard platforms.
// Other platforms route events differently, e.g. on AndroidTV
// focus is handled natively... (same text you have).
#endif
```

File: cobalt/shell/browser/shell_test_support.h (hunk 4 above)
Comment (yell0wd0g) [inline cobalt/shell/browser/shell_test_support.h]: Is this change intended? Usually file copyright years never change.

File: cobalt/browser/cobalt_content_browser_client.cc [hunk 7]
```diff
@@ -161,12 +161,16 @@ CobaltContentBrowserClient::CobaltContentBrowserClient(
...
+  ui::PlatformWindowStarboard::SetWindowDestroyedCallback(
+      base::BindRepeating(&CobaltContentBrowserClient::OnSbWindowDestroyed,
+                          weak_factory_.GetWeakPtr()));
 #endif  // BUILDFLAG(IS_STARBOARD)
 }
 
 CobaltContentBrowserClient::~CobaltContentBrowserClient() {
 #if BUILDFLAG(IS_STARBOARD)
   ui::PlatformWindowStarboard::SetWindowCreatedCallback(base::NullCallback());
+  ui::PlatformWindowStarboard::SetWindowDestroyedCallback(base::NullCallback());
```
Comment (yell0wd0g) [inline cobalt/browser/cobalt_content_browser_client.cc]: Just double checking: a NullCallback will crash (DCHECK) if
called (unless is tested, which is rarely done). Is this 
intended here? If the intention is to allow for such 
callbacks but do nothing, then `base::DoNothing()` is the 
appropriate callback.

File: cobalt/shell/browser/shell.cc [hunk 8]
```diff
@@ -322,6 +365,12 @@ void Shell::Shutdown() {
     return;
   }
 
+  static bool is_shutting_down = false;
```
Comment (yell0wd0g) [inline cobalt/shell/browser/shell.cc]: This is to protect the critical section in l.374-395
from reentrancy, right? If so then I'd rather use
concurrency specific patterns like a homebrew:

```
  std::atomic_flag is_shutting_down = ATOMIC_FLAG_INIT;;
  if (is_shutting_down.test_and_set()) {
    return;
  }
  // ...
  is_shutting_down.clear();
```

Or a `base::Lock` ([code](https://source.corp.google.com/h/lbshell-internal/cobalt_src/+/main:base/synchronization/lock.h;l=30;drc=21fb7d148f9e0948484b16d75a2467ef6c4b0787)).

File: cobalt/shell/browser/shell_platform_delegate.cc [hunk 9]
```diff
@@ -25,16 +27,64 @@ bool ShellPlatformDelegate::IsVisible() const {
...
+    const gfx::Size& default_window_size,
+    bool is_visible) {
+  is_visible_ = is_visible;
+}
+
+void ShellPlatformDelegate::OnBlur() {
+  for (auto* shell : Shell::windows()) {
```
Comment (yell0wd0g) [inline cobalt/shell/browser/shell_platform_delegate.cc:32]: Question: Under what condition would Cobalt have multiple 
`windows()`? If none can be conjured then consider leaving 
a `CHECK_EQ(Shell::windows().size(), 1u)` and removing the
`for` loop. Same comment in the other methods below.

Note that `this` has a `web_contents()` and a
`splash_screen_web_contents()` IIUC.

File: cobalt/shell/browser/shell_platform_delegate.cc [hunk 10]
```diff
@@ -25,16 +27,64 @@ bool ShellPlatformDelegate::IsVisible() const {
...
+}
+
+void ShellPlatformDelegate::OnFocus() {
+  for (auto* shell : Shell::windows()) {
+    shell->web_contents()->Focus();
+  }
+}
+
+void ShellPlatformDelegate::OnConceal() {
+  if (!IsVisible()) {
```
Comment (yell0wd0g) [inline cobalt/shell/browser/shell_platform_delegate.cc]: Question: is it possible to receive an `OnConceal()` while
not-visible? Same question in l.60. If there is a specific
use case consider leaving a comment, otherwise maybe a 
`CHECK(IsVisible());`?

File: cobalt/shell/browser/shell_platform_delegate.cc [hunk 11]
```diff
@@ -25,16 +27,64 @@ bool ShellPlatformDelegate::IsVisible() const {
   return is_visible_;
 }
 
+__attribute__((weak)) void ShellPlatformDelegate::Initialize(
+    const gfx::Size& default_window_size,
```
Comment (yell0wd0g) [inline cobalt/shell/browser/shell_platform_delegate.cc]: micro-nit: In Chromium it's common to escape unused params, i.e.
` const gfx::Size& /*default_window_size*/,` 🤷

File: cobalt/shell/browser/shell_platform_delegate_aura.cc [hunk 12]
```diff
@@ -111,6 +107,13 @@ void ShellPlatformDelegate::RevealShell(Shell* shell) {
 
   platform_->aura->host()->Show();
 }
+
+void ShellPlatformDelegate::ConcealShell(Shell* shell) {
+  if (platform_->aura) {
```
Comment (yell0wd0g) [inline cobalt/shell/browser/shell_platform_delegate_aura.cc]: Is this possible? Maybe if we start the App "minimized"? A 
comment on this would be super helpful for future developers,
e.g.
```
  // This could happen if we start minimized or not visible,
  // e.g. in pre-app configuration
  if (...
```

File: cobalt/app/app_lifecycle_delegate.cc (hunk 6 above)
Comment (jellefoks) [reply cobalt/app/app_lifecycle_delegate.cc]: Done

File: cobalt/shell/browser/shell_test_support.h (hunk 4 above)
Comment (jellefoks) [reply cobalt/shell/browser/shell_test_support.h]: Oops. Done.

File: cobalt/browser/cobalt_content_browser_client.cc (hunk 7 above)
Comment (jellefoks) [reply cobalt/browser/cobalt_content_browser_client.cc]: They can also be null of SetWindowFooCallback is never called, so the code in platform_window_starboard.cc  referencing the callback already guards against it being null:

```
  if (g_created_callback) {
    (*g_created_callback).Run(sb_window_);
  }
```

But, I'll change the static initializers to `base::DoNothing()`, because that's cleaner. Needs #include "base/functional/callback_helpers.h"

File: cobalt/shell/browser/shell.cc (hunk 8 above)
Comment (jellefoks) [reply cobalt/shell/browser/shell.cc]: This is actually superfluous and was from an earlier attempt to solve a bug that was elsewhere. The g_platform check above covers this. Note: since this is always called from the same thread, it doesn't have to be atomic.

File: cobalt/shell/browser/shell_platform_delegate.cc (hunk 11 above)
Comment (jellefoks) [reply cobalt/shell/browser/shell_platform_delegate.cc]: Done.

File: cobalt/shell/browser/shell_platform_delegate.cc (hunk 9 above)
Comment (jellefoks) [reply cobalt/shell/browser/shell_platform_delegate.cc:32]: The fact that `Shell` in `cobalt/shell/browser/shell.h` has support for multiple windows with its `static std::vector<Shell*> windows` means that we should carry that through here. 

We could remove support for multiple windows altogether, but doing that would span outside of the scope of this PR. Filed as b/491969758. I do not know if there currently is an edge case or test setup where there is an additional window.

File: cobalt/shell/browser/shell_platform_delegate.cc (hunk 10 above)
Comment (jellefoks) [reply cobalt/shell/browser/shell_platform_delegate.cc]: This mirrored the `if(visible_)` check in OnReveal added for a `LifecycleTest::RedundantReveal` test, but that is the only time it can actually occur in production.

So, good point, removed that test and changed to a CHECK() here and in the other state change functions. A followup PR is already prepared that ensures it will never happen redundantly.

File: cobalt/shell/browser/shell_platform_delegate_aura.cc (hunk 12 above)
Comment (jellefoks) [reply cobalt/shell/browser/shell_platform_delegate_aura.cc]: No, is not possible. Was a leftover defensive change. Removed

File: cobalt/app/app_lifecycle_delegate.cc (hunk 6 above)
Comment (jellefoks) [reply cobalt/app/app_lifecycle_delegate.cc]: This doesn't build on `android-arm64 / arm64_devel` and `android-arm / arm_devel`...

```
2026-03-12T07:39:28.6071814Z #error This file is only intended for Starboard platforms.
2026-03-12T07:39:28.6072446Z  ^
2026-03-12T07:39:28.6073107Z ../../cobalt/app/app_lifecycle_delegate.cc:155:11: error: use of undeclared identifier 'platform_event_source_'
2026-03-12T07:39:28.6073944Z       if (platform_event_source_) {
2026-03-12T07:39:28.6074336Z           ^
2026-03-12T07:39:28.6075066Z ../../cobalt/app/app_lifecycle_delegate.cc:156:9: error: use of undeclared identifier 'platform_event_source_'
2026-03-12T07:39:28.6075955Z         platform_event_source_->HandleFocusEvent(event);
2026-03-12T07:39:28.6076445Z         ^
2026-03-12T07:39:28.6076725Z 3 errors generated.
```

File: ui/ozone/platform/starboard/platform_window_starboard.cc [hunk 13]
```diff
@@ -136,20 +150,52 @@ gfx::Rect PlatformWindowStarboard::GetBoundsInDIP() const {
...
+    widget_available_ = true;
+    delegate_->OnAcceleratedWidgetAvailable(
+        reinterpret_cast<intptr_t>(SbWindowGetPlatformHandle(sb_window_)));
+  }
 }
 
 void PlatformWindowStarboard::Hide() {
-  NOTIMPLEMENTED_LOG_ONCE();
+  VLOG(1) << "PlatformWindowStarboard::Hide";
+  if (!SbWindowIsValid(sb_window_)) {
```
Comment (andrewsavage1) [inline ui/ozone/platform/starboard/platform_window_starboard.cc]: When would this case be triggered?

File: ui/ozone/platform/starboard/platform_window_starboard.h [hunk 14]
```diff
@@ -16,6 +16,7 @@
 #define UI_OZONE_PLATFORM_STARBOARD_PLATFORM_WINDOW_STARBOARD_H_
 
 #include "base/functional/callback.h"
+#include "base/memory/raw_ptr.h"
```
Comment (andrewsavage1) [inline ui/ozone/platform/starboard/platform_window_starboard.h:19]: What is this needed for?

File: ui/views/widget/desktop_aura/desktop_window_tree_host_platform.cc [hunk 15]
```diff
@@ -881,6 +881,11 @@ void DesktopWindowTreeHostPlatform::OnWillDestroyAcceleratedWidget() {
   desktop_native_widget_aura_->OnHostWillClose();
 }
 
+void DesktopWindowTreeHostPlatform::OnAcceleratedWidgetDestroyed() {
```
Comment (andrewsavage1) [inline ui/views/widget/desktop_aura/desktop_window_tree_host_platform.cc:885]: Should be guarded by buildflags

File: ui/views/widget/desktop_aura/desktop_window_tree_host_platform.h [hunk 16]
```diff
@@ -152,6 +152,7 @@ class VIEWS_EXPORT DesktopWindowTreeHostPlatform
   void OnCloseRequest() override;
   void OnAcceleratedWidgetAvailable(gfx::AcceleratedWidget widget) override;
   void OnWillDestroyAcceleratedWidget() override;
+  void OnAcceleratedWidgetDestroyed() override;
```
Comment (andrewsavage1) [inline ui/views/widget/desktop_aura/desktop_window_tree_host_platform.h:156]: Should be guarded by buildflags

File: cobalt/shell/browser/shell_platform_delegate.cc (hunk 3 above)
Comment (jellefoks) [reply cobalt/shell/browser/shell_platform_delegate.cc]: Yeah, this is not needed anymore this was the 'base class' implementation added before the platform specializations were added to the non-derived classes in `shell_platform_delegate_(aura|views|android|ios).(cc|mm)`. Same for `RevealShell` and `ConcealShell` below.

File: cobalt/app/app_lifecycle_delegate.cc (hunk 6 above)
Comment (yell0wd0g) [reply cobalt/app/app_lifecycle_delegate.cc]: Hmm so it is being built. While looking at who pulls from
it I noticed this is 26.eap, which doesn't make sense for
AndroidTV... Should we leave the current change (with `#error`)
and remove the file from android builds?

File: cobalt/app/app_lifecycle_delegate.cc (hunk 2 above)
Comment (jellefoks) [reply cobalt/app/app_lifecycle_delegate.cc]: Addressed  in the 3rd commit of this PR.

File: cobalt/app/app_lifecycle_delegate.cc (hunk 6 above)
Comment (jellefoks) [reply cobalt/app/app_lifecycle_delegate.cc]: I'll address it with an 'if (is_starboard)' build gate in a PR update.

File: ui/ozone/platform/starboard/platform_window_starboard.cc (hunk 13 above)
Comment (jellefoks) [reply ui/ozone/platform/starboard/platform_window_starboard.cc]: Only if the platform sends duplicate conceal events, which I had a test for here but I'll prevent from propagating in #9424 (nearly ready, just needs rebasing on this PR). 

Added `CHECK()` instead, and for `widget_available_` here and in `Show()`.

File: ui/ozone/platform/starboard/platform_window_starboard.h (hunk 14 above)
Comment (jellefoks) [reply ui/ozone/platform/starboard/platform_window_starboard.h:19]: Style rule "include what you use". It's an addition for a missing  for the raw_ptr<> on line 103 below.

https://google.github.io/styleguide/cppguide.html#Include_What_You_Use

File: ui/views/widget/desktop_aura/desktop_window_tree_host_platform.cc (hunk 15 above)
Comment (jellefoks) [reply ui/views/widget/desktop_aura/desktop_window_tree_host_platform.cc:885]: Done

File: ui/views/widget/desktop_aura/desktop_window_tree_host_platform.h (hunk 16 above)
Comment (jellefoks) [reply ui/views/widget/desktop_aura/desktop_window_tree_host_platform.h:156]: Done

File: starboard/shared/x11/application_x11.cc (hunk 5 above)
Comment (jellefoks) [reply starboard/shared/x11/application_x11.cc]: I had encountered a crash here dereferencing the window. But auditing this, that should only be possible if DestroyWindow can be called from a diffent thread, which if that happens should be addressed differently. I changed this back but added SB_CHECK() to ensure it asserts instead of undefined behavior while dereferencing.

File: cobalt/BUILD.gn [hunk 17]
```diff
@@ -109,14 +109,14 @@ source_set("common") {
...
-    "app/app_lifecycle_delegate.cc",
-    "app/app_lifecycle_delegate.h",
     "app/cobalt_main_delegate.cc",
     "app/cobalt_main_delegate.h",
   ]
 
   if (is_starboard) {
     sources += [
+      "app/app_lifecycle_delegate.cc",
+      "app/app_lifecycle_delegate.h",
```
Comment (yell0wd0g) [inline cobalt/BUILD.gn:119]: \o/

File: ui/ozone/platform/starboard/platform_window_starboard.cc (hunk 13 above)
Comment (jellefoks) [reply ui/ozone/platform/starboard/platform_window_starboard.cc]: Actually, in `Show()` the check is needed due to a convoluted initialization process around the various abstractions and the splash screen that I have yet to unravel.

File: cobalt/app/app_lifecycle_delegate.cc [hunk 18]
```diff
@@ -146,7 +144,20 @@ void AppLifecycleDelegate::HandleEvent(const SbEvent* event) {
       OnStop(event);
       break;
     case kSbEventTypeBlur:
+      content::Shell::OnBlur();
+#if !BUILDFLAG(IS_STARBOARD)
+#error This file is only intended for Starboard platforms.
```
Comment (andrewsavage1) [inline cobalt/app/app_lifecycle_delegate.cc:149]: It's a bit of a strange place to put this error message, I would expect it at the top of the file—I don't think it's important enough to change now, but let's do it when we're cherry picking to main

**Review by andrewsavage1 — APPROVED** [review]

**Comment by andrewsavage1** [comment]
Look great!
